### DISCUSSION TITLE
Saleor Core & Frontend on different domains

### DISCUSSION THREAD

**Discussion by JulianAtTheFrontend** [saleor/saleor#8470] [Q&A]
Is it generally possible to host core and frontend on different domains? I'm currently running into trouble with saving httpOnly response cookies as it's a cross site request. Even hosting the core on saleor.mydomain.com and the frontend on mydomain.com still gives me a headache. 

Not sure what requirements need to be met in order to save that cookie.

**Comment by manuelaguirre** [comment]
@JulianAtTheFrontend I'm having the SAME issue, did you fix it?

**Reply by JulianAtTheFrontend** [reply]
Unfortunately not. Still trying to figure out what causes the error. Postet on Stackoverflow but haven't found a solution yet. 

My Cookie from the response is secure, httpOnly & environment vars are all set. According to some documentations in the web this should work out. A recent update of Saleor also applied the SameSite : None attribute - however, still no progress.

For development I was proxying my requests but for production this won't really work 🤷‍♂️

Kind of stuck right now but I'm sure this is supposed to work out somehow (otherwise a headless ecommerce wouldn't make sense if it weren't headless). 

If you find any solution please let me know ✌️✌️

**Reply by manuelaguirre** [reply]
I will thank you

**Comment by patrys** [comment]
@JulianAtTheFrontend What exact issue are you facing? You should not have to save any cookies as the only cookies set by Saleor are the refresh tokens that are not meant to be seen by your storefront. It's intentionally only visible to the API and should be configured with a security policy (`SameSite=None`, this was implemented in Saleor 3.0) that automatically includes it for all cross-origin requests.

**Reply by JulianAtTheFrontend** [reply]
Yes, it's about the refresh cookies. The issue is that these cookies won't get set. E.g. after logging in and manually refreshing the page there is no (cookie-) refresh token to verify the user, so the user needs to login again.

I initially thought this issue will be fixed with the recent sameSite=None update, but it's still not setting the cookie.

I am aware of that this cookie is httpOnly and can not be changed or viewed by the storefront, however I can see if it is set within the browsers dev tools.

It's just hard to debug when there is no error.

![image](https://user-images.githubusercontent.com/92525148/140637099-6d1e372a-5b67-4ad4-9707-fe8ee6b33c41.png)

![image](https://user-images.githubusercontent.com/92525148/140637097-079f6259-3174-428e-86df-8d986590e413.png)

![image](https://user-images.githubusercontent.com/92525148/140637090-d39a3d95-29d8-485b-bf3e-80685c17f849.png)

This is the cookie response I'm receiving. But this cookie won't get set.

It was working during development while proxying the requests (so the domain equals saleor core domain), which now is of course different.

**Reply by patrys** [reply]
@jwm0, any idea what's happening here? Could it be the browser rejecting third-party cookies?

**Reply by JulianAtTheFrontend** [reply]
Any news about this?

**Reply by jwm0** [reply]
Are there any console warning/errors? It looks like the cookie is properly sent from the server with correct `SameSite=None`, are you sure it's not your browser blocking these cookies?

**Reply by JulianAtTheFrontend** [reply]
No errors or warnings. Yes, it looks like to cookie is properly sent, however it won't be saved in any browser or device. What is the "domain" field supposed to contain? 

I was thinking that there might be an error with setting the environment vars on Heroku - but I can't tell if there's an error somewhere or something that is not working as it is should be.

**Comment by jbergstroem** [comment]
We are moving towards a similar setup (`saleor.mysite.com` and `mysite.com`) and my guess is that since saleor sets the domain `saleor.mysite.com` for the refresh cookie, ~it _could_ be seen as a 3rd party cookie since the first cookie set is `mysite.com`~. Found a similar answer on stackoverflow: https://stackoverflow.com/questions/69665080/cookie-for-a-refresh-token-is-not-stored ~- that suggests this pattern but has similar issues~.

Also, I think that moving to `samesite=Strict` would make sense for most setups, no? 

Edit: I redact my comment; after testing the conclusion is that it works.

**Reply by JulianAtTheFrontend** [reply]
@jbergstroem how exactly did you make it work? I also tested again to host Saleor Core on `saleor.mysite.com` plus changing the cookie domain to `.mysite.com` and `samesite=Strict` in saleor middleware.py. Unfortunately, still no success.

**Reply by PlusA2M** [reply]
The solution is to set `DEBUG=False` for the environment variable therefore the cookies' `SameSite=None` and `Secure` is set. For the development convenience you can create a `docker-compose.debug.yml` to set the `JWT_TTL_ACCESS` to something like `1d` or `2h` at the development stage if you need `DEBUG=True`.

**Reply by TurkaljGoran** [reply]
We had the same issues on our projects.
Setting DEBUG=False solved the refresh token issue since the http only cookie worked as intended, BUT - media files were no longer loaded. 

We then needed to transfer all media files to Google Cloud Platform / Storage and serve them from there.

**Reply by jbergstroem** [reply]
This is how our setup loooks (foo.com for our sveltekit frontend, gql.foo.com for saleor, files.foo.com for our s3-like storage minio, dash.foo.com for saleor dashboard):
1. We use caddy as a proxy traffic from internet that handles https termination. Here's our configuration - `SALEOR_URL` and `SALEOR_HOST` are environment variables (same config for production/staging). Admin is used for prometheus export.
```
{
	auto_https off
	https_port 8080
	admin :2019
}

handle_errors {
	header -server
}

:8080 {
	header -server
	tls /etc/caddy/cert.pem /etc/caddy/cert.key {
		protocols tls1.3
		resolvers 1.1.1.1
	}
	encode gzip zstd
	reverse_proxy {$SALEOR_URL} {
		header_up Host "{$SALEOR_HOST}"
		header_up X-Forwarded-Proto "https"
		header_down -server
	}
	log {
		level INFO
	}
}
```
2. Saleor uses:
    - `DEBUG=0`
    - `ENABLE_SSL=1`
    - `ALLOWED_CLIENT_HOSTS=foo.com, dash.foo.com, gql.foo.com` (last for playground)
    - `ALLOWED_HOSTS=gql.foo.com`
    - `ALLOWED_GRAPHQL_ORIGINS=https://foo.com, https://gql.foo.com, https://foo.com, https://dash.foo.com`
    - `AWS_MEDIA_BUCKET_NAME=static`
    - `AWS_MEDIA_CUSTOM_DOMAIN=files.foo.com/static`
    - `AWS_DEFAULT_ACL=public-read`
    - `AWS_S3_ENDPOINT_URL=http://foo-minio.internal:9000` (only accessible on internal network)

The "magic" is really about understanding how `ALLOWED_CLIENT_HOSTS` and `ALLOWED_GRAPHQL_ORIGINS` works.
